TZ12 feat: Develop

enjsite opened the pull request.

Tulpanchik2906 committed:
feat: первый вариант функциональности Отзывы.

Tulpanchik2906 committed:
fix: checkstyle

Tulpanchik2906 committed:
fix: checkstyle

Tulpanchik2906 committed:
feat: добавлена функциональность рекомендаций

Tulpanchik2906 committed:
feat: добавлена функциональность рекомендаций

Tulpanchik2906 committed:
feat: добавлено логирование

cptntotoro committed:
feat: добавить режиссёров

enjsite committed:
Merge pull request #8 from enjsite/add-reviews

feat: первый вариант функциональности Отзывы.

Tulpanchik2906 committed:
merge develop

enjsite committed:
Merge pull request #11 from enjsite/add-recommendations

feat: добавлена функциональность рекомендаций

cptntotoro committed:
fix: доработать предложенные исправления

stgmakarov committed:
fit: Добавлена функциональность общих фильмов

enjsite committed:
Merge pull request #13 from enjsite/add-director

feat: добавить режиссёров

stgmakarov committed:
fix: переносы строк и прочее (без изм. функционала)

stgmakarov committed:
Merge branch 'develop' into add-common-films

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java

stgmakarov committed:
Merge pull request #14 from enjsite/add-common-films

fit: Добавлена функциональность общих фильмов

GeiseR98 committed:
feat: переписал 1 заходом

enjsite committed:
Merge pull request #10 from enjsite/add-remove-endpoint

Add remove endpoint

GeiseR98 committed:
add-search

enjsite committed:
feature: TZ12, user feed

enjsite committed:
feature: TZ12, user feed, fix del db

enjsite committed:
feature: TZ12, user feed, fix checkstyle

enjsite committed:
Merge pull request #15 from enjsite/add-feed

feature: TZ12, user feed

stgmakarov committed:
fit: Добавить возможность выводить топ-N фильмов по количеству лайков.

stgmakarov committed:
fix: пробелы

stgmakarov committed:
fix: добавление вариаций, когда есть года нет жанра и наоборот

enjsite committed:
Merge pull request #16 from enjsite/add-search

add-search

stgmakarov committed:
Merge remote-tracking branch 'origin/develop' into add-most-populars

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java

enjsite committed:
Merge pull request #17 from enjsite/add-most-populars

fit: Добавить возможность выводить топ-N фильмов по количеству лайков.

enjsite committed:
feature: TZ12, user feed, fix for removing users

zgreen3 commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -188,4 +323,108 @@ public void addLike(Integer filmId, Integer userId) {
+                "WHERE f.ID IN\n" +
+                "(SELECT film_id FROM LIKES\n" +
+                "WHERE USER_ID = ? AND\n" +
+                "FILM_ID IN ( SELECT FILM_ID FROM LIKES WHERE USER_ID = ? ))\n" +
```
Надо исправить:
Могу ошибаться (не было возможности запустить проект с БД для проверки), но здесь вложенный запрос 2-го уровня выглядит излишним: отбор фильмов по ИД пользователя уже производится в таблице лайков на сточку выше, на 387 строке. 
Проверьте, пожалуйста, этот момент и удалите лишний запрос (всю данную строчку 388 из фильтра внешнего к данному select-запроса) или аргументируйте необходимость его наличия здесь.
Комментарий указал как "Надо исправить", потому что в бэкенд-проектах работа с БД обычно очень чувствительная для производительности всего приложения тема и ей всегда важно уделять повышенное внимание=).

Tulpanchik2906 replied:
Так сделано специально, иначе у меня терялись пустые строки

zgreen3 commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -188,4 +323,108 @@ public void addLike(Integer filmId, Integer userId) {
+                "                     ORDER BY cnt DESC) AS common_likes ON u.ID = common_likes.user_id\n" +
+                "                  GROUP BY u.id\n" +
+                "                  ORDER BY cnt2 DESC) AS user_cnt))\n" +
+                "     EXCEPT SELECT film_id\n" +
```
Можно лучше:
Обычно в движки реляционных СУБД оптимизированы для работы с JOIN-ами, а выражения по работе с множествами значений, вроде EXCEPT SELECT менее распросранены и могут выдавать неоптимальный результат на разных СУБД (особенно, если они применяются уже в конце реализации плана запроса), кроме того, иногда они сложнее воспринимаются читающим код, т.к. мысленно их нужно выполнять над мысленным результатом работы всего остального кода, а не в самом этом процессе.
Подумайте, как можно заменить EXCEPT SELECT (подсказка: на LEFT OUTER JOIN с фильтром на is null у соответствующей таблицы / подзапроса, см. пример этого, например, здесь, на англ. яз., https://stackoverflow.com/a/56859428).

zgreen3 commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,86 @@
+        log.info("Получен запрос на добавление лайка" +
+                " для отзыва с id: {} от пользователя с id: {} ", id, userId);
+        reviewService.addLike(id, userId);
+        return "Лайк добавлен";
```
Можно лучше:
Здесь и далее в коде целесообразно возвращать сообщение о результате работы вызываемого метода сервиса (для чего, вероятно, его потребуется немного преобразовать для возврата булевых значений), т.к. в REST-эндпоинтах (даже с типом POST, PUT и DELETE) на этапе работы методов в сервисах и репозиториях не исключены всевозможные сбои в работе и клиенту важно понимать, если они имели место, т.е. не получать в таких случаях одно и то же сообщение о том, что нужное действие сервером произведено.

zgreen3 commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
Можно лучше:
Лишний закомментированный код (в этом файле и в файле UserControllerTest) лучше удалять даже в тестовых методах при пулл-реквестах в мастер ветку и вообще, в любые "общие" (не фичи) ветки, где работает более 1 разработчика, т.к. иногда он может совершенно неожиданно "выстрелить в ногу" (пишу на собственном подобном опыте)), если попадает в распоряжение к другим разработчикам.

zgreen3 commented on src/main/java/ru/yandex/practicum/filmorate/storage/UserDbStorage.java:
```diff
@@ -17,6 +21,11 @@ public class UserDbStorage implements UserStorage {
     public User get(Integer id) {
 
+        String checkQuery = "SELECT COUNT(*) FROM users WHERE id = ?";
+        if (jdbcTemplate.queryForObject(checkQuery, Integer.class, id) == 0) {
```
Можно лучше:
Здесь важно добавить проверку результата на null до проверки на 0, т.к. возвращаемое значение этого метода будет классом Integer, который может быть null (см., например, пояснение на англ. яз., https://stackoverflow.com/a/74433112).

zgreen3 reviewed (changes requested):
Привет! Вы проделали большую работу! 
У вас получился отличный проект и появились навыки работы в команде а также перекрёстных ревью кода внутри команды (насколько могу судить из истории пулл-реквестов, это была также плодотворная часть вашей работы)). 
Молодцы, что совместно сделали практически все возможные задания (поскольку не нашёл описания общего пулл-реквеста в ветку main и данных об этом в файле readme.md проекта, судил об этом по названию веток, комментариям в пулл-реквестах в ветку develop и по коду самого проекта - в будущих командных проектах важно описывать их функционал и структуру, хотя бы парой фраз, как и редкие технические решения, при их наличии, в файле readme.md - никогда не знаешь, кому передадут проект "по наследству", а такой файл может снять сразу множество вопросов)!
Вы хорошо разобрались во взаимодействие между слоями "контроллер", "сервис", "репозиторий" и в том, как добавлять новый функционал в эту трёхзвенную архитектуру.  
У вас получились отличные запросы к таблицам (с использованием вложенных запросов, но не более, чем на 1 уровень в глубину, и даже с применением аналитических функций), хотя возможны отдельные оптимизации (какие именно я указал в комментариях в гите). 
Также хочу заметить, что к тестовым классам важно относиться как к полноценным классам проекта, т.е. не пренебрегать стилистическими деталями (например, удалять неиспользуемый код при пулл-реквестах в общие ветки проекта, определять оптимальные в каждом конкретном случае области видимости полей, и т.п., это поможет не вносить в юнит-тесты случайные трудноуловимые ошибки в будущем). 
Обращаю внимание, что в REST-эндпоинтах (даже с типом POST, PUT и DELETE) с синхронной обработкой (асинхрную вы пока не проходили, поэтому пока можно считать, что во всех)) желательно возвращать не константный результат, а полноценное сообщение о статусе ответа, т.к. на этапе работы методов в сервисах и репозиториях не исключены всевозможные сбои в работе и клиенту важно понимать, если они имели место, т.е. не получать в таких случаях одно и то же сообщение о том, что нужное действие сервером произведено.
После учёта комментариев и (или) возможных изменений в коде на их основе, присылайте работу на завершение проверки.
Спасибо.

enjsite committed:
feature: TZ12, del comments

Tulpanchik2906 committed:
fix: 5-ый коммент от ревьювера Можно лучше:
                               Здесь важно добавить проверку результата на null до проверки на 0, т.к. возвращаемое значение этого метода будет классом Integer, который может быть null (см., например, пояснение на англ. яз., https://stackoverflow.com/a/74433112).

Tulpanchik2906 committed:
fix: 5-ый коммент от ревьювера check-style

Tulpanchik2906 committed:
fix: 3-ый коммент от ревьювера: Можно лучше:
                                Здесь и далее в коде целесообразно возвращать сообщение о результате работы вызываемого метода сервиса (для чего, вероятно, его потребуется немного преобразовать для возврата булевых значений), т.к. в REST-эндпоинтах (даже с типом POST, PUT и DELETE) на этапе работы методов в сервисах и репозиториях не исключены всевозможные сбои в работе и клиенту важно понимать, если они имели место, т.е. не получать в таких случаях одно и то же сообщение о том, что нужное действие сервером произведено.

Tulpanchik2906 committed:
fix: ревью 4-ый комментарий: Можно лучше:
                             Лишний закомментированный код (в этом файле и в файле UserControllerTest) лучше удалять даже в тестовых методах при пулл-реквестах в мастер ветку и вообще, в любые "общие" (не фичи) ветки, где работает более 1 разработчика, т.к. иногда он может совершенно неожиданно "выстрелить в ногу" (пишу на собственном подобном опыте)), если попадает в распоряжение к другим разработчикам.

Tulpanchik2906 committed:
fix: 1-ый коммент ревьювера (вместо подзапроса и объединения логика вынесена сразу в блок where)
Надо исправить:
Могу ошибаться (не было возможности запустить проект с БД для проверки), но здесь вложенный запрос 2-го уровня выглядит излишним: отбор фильмов по ИД пользователя уже производится в таблице лайков на сточку выше, на 387 строке.
Проверьте, пожалуйста, этот момент и удалите лишний запрос (всю данную строчку 388 из фильтра внешнего к данному select-запроса) или аргументируйте необходимость его наличия здесь.
Комментарий указал как "Надо исправить", потому что в бэкенд-проектах работа с БД обычно очень чувствительная для производительности всего приложения тема и ей всегда важно уделять повышенное внимание=).

Tulpanchik2906 committed:
fix: 2-ой коммент ревьювера замена except на not in
Можно лучше:
Обычно в движки реляционных СУБД оптимизированы для работы с JOIN-ами, а выражения по работе с множествами значений, вроде EXCEPT SELECT менее распросранены и могут выдавать неоптимальный результат на разных СУБД (особенно, если они применяются уже в конце реализации плана запроса), кроме того, иногда они сложнее воспринимаются читающим код, т.к. мысленно их нужно выполнять над мысленным результатом работы всего остального кода, а не в самом этом процессе.
Подумайте, как можно заменить EXCEPT SELECT (подсказка: на LEFT OUTER JOIN с фильтром на is null у соответствующей таблицы / подзапроса, см. пример этого, например, здесь, на англ. яз., https://stackoverflow.com/a/56859428).

Tulpanchik2906 committed:
fix: оптимизация запроса общих фильма, убран джоин к таблице films

zgreen3 commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -379,14 +380,15 @@ public List<Film> getRecommendations(Integer id) {
-                "WHERE USER_ID = ? AND\n" +
-                "FILM_ID IN ( SELECT FILM_ID FROM LIKES WHERE USER_ID = ? ))\n" +
-                "ORDER BY LIKESCOUNT.cnt DESC";
+        String sqlReq = "SELECT FILM_ID AS ID\n" +
```
Отлично (и немного "Можно лучше"):
Вижу, что запрос упростился: убран left join и данные об ИД фильмов теперь получаются напрямую из таблицы лайков, т.к. возможно считать лайки через подсчёт числа пользователей (их ИД) для каждого из фильмов в таблице лайков, на что я не обратил внимание при предыдущей итерации ревью - вы сами это заметили и оптимизировали, круто!

При этом в оптимизированном коде остался подзапрос
```
"  AND (FILM_ID in\n" +
"         (SELECT FILM_ID\n" +
"          FROM LIKES\n" +
"          WHERE USER_ID = ?))\n" + 
```
, необходимость наличия которого для меня всё ещё не очевидна. 
При предыдущей итерации код-ревью я не обратил в этом месте ваше внимание (сам тогда про это не подумал, сейчас исправляюсь)), что нужно попробовать не просто удалить этот подзапрос, а заменить его на фильтр `user_id = ... AND film_id IS NOT NULL`, чтобы получить эквивалентный (только в данном конкретном select-е!) выражению `user_id = ... AND film_id IN (...)` фильтр, который не будет возвращать строки, где film_id может быть null. В любом случае, этот вопрос не настолько важен для обучения, чтобы конкретно из-за него проводить ещё одну итерацию код-ревью =().

zgreen3 commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -379,14 +380,15 @@ public List<Film> getRecommendations(Integer id) {
-                "ORDER BY LIKESCOUNT.cnt DESC";
+        String sqlReq = "SELECT FILM_ID AS ID\n" +
+                "FROM LIKES\n" +
+                "WHERE USER_ID = ?\n" +
```
Можно лучше:
В синтаксисе sql-запросов обычно принято сами команды (ключевые слова языка SQL) писать заглавными буквами, чтобы их было видно во всём запросе, а имена переменных - прописными.

zgreen3 commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -45,42 +45,38 @@ public Review update(@Valid @RequestBody Review review) throws ValidationExcepti
 
     @DeleteMapping("/{id}")
-    public String deleteById(@PathVariable Integer id) {
+    public boolean deleteById(@PathVariable Integer id) {
```
Можно лучше:
Отлично, что вы изменили возвращаемые значения на уровне сервиса и репозитория!

На уровне контроллера можно (и даже желательно, т.к. API могут использовать клиенты, которые не в курсе того, что будет означать возвращаемое булево значение) оставить возврат строк / Stringв тело http-ответов (т.е. как возвращаемые значения методов контроллера в Spring Boot), просто текст возвращаемого сообщения можно сделать разным - в зависимости от булева значения из метода сервиса (успех или ошибка).

zgreen3 reviewed (approved):
Привет, вижу, что вы учли все важные комментарии, отлично (вижу по истории коммитов, что вчера вечером и даже ночью было приложено достаточно сил для этого, такая целеустремлённость вызывает уважение)!
Принимаю проект в текущем виде, встретимся на проверке следующих проектов.
Успехов в обучении!
